### PR TITLE
[NFC][SYCL] Move check_type_in_v to sycl/detail/type_traits.hpp

### PR DESCRIPTION
**Pull request by aelovikov-intel**
It's generic enough and isn't specific to builtins.

### CODE REVIEW

File: sycl/include/sycl/detail/type_traits.hpp [hunk 1]
```diff
@@ -532,6 +532,20 @@ struct map_type<T, From, To, Rest...> {
   using type = std::conditional_t<std::is_same_v<From, T>, To,
                                   typename map_type<T, Rest...>::type>;
 };
+template <typename T, typename... Ts> constexpr bool CheckTypeIn() {
+  constexpr bool SameType[] = {
+      std::is_same_v<std::remove_cv_t<T>, std::remove_cv_t<Ts>>...};
+  // Replace with std::any_of with C++20.
```
Comment (maarquitos14) [inline sycl/include/sycl/detail/type_traits.hpp:538]: Now that we are looking into this, does it make sense to do what this comment suggests? Maybe in a different PR?

Not a blocking comment, though.

**Review by maarquitos14 — APPROVED** [review]
LGTM.

File: sycl/include/sycl/detail/type_traits.hpp (hunk 1 above)
Comment (aelovikov-intel) [reply sycl/include/sycl/detail/type_traits.hpp:538]: We're at C++17, unfortunately...

**Comment by aelovikov-intel** [comment]
Post-commit failed in https://github.com/intel/llvm/actions/runs/7614418709/job/20737377227 with:

```
env ONEAPI_DEVICE_SELECTOR=level_zero:gpu  /__w/llvm/llvm/build-e2e/ESIMD/Output/matrix_transpose_glb.cpp.tmp.out
# executed command: env ONEAPI_DEVICE_SELECTOR=level_zero:gpu /__w/llvm/llvm/build-e2e/ESIMD/Output/matrix_transpose_glb.cpp.tmp.out
# .---command stdout------------
# | Number of iterations: 10
# | [KernelTime]: 50.959
# | [OverallTime][Primary]: 96.6882
# `-----------------------------
# .---command stderr------------
# | 
# | Transpose square matrix of size 32
# | Error at M(0, 18) = 576
# | 
# | Transpose square matrix of size 512
# | 
# | Transpose square matrix of size 1024
# | 
# | Transpose square matrix of size 2048
# | 
# | Transpose square matrix of size 4096
# | 
# | Transpose square matrix of size 512
# | Error at M(0, 18) = 9216
# | 
# | Transpose square matrix of size 1024
# | Error at M(0, 18) = 18432
# | 
# | Transpose square matrix of size 2048
# | Error at M(0, 18) = 36864
# | 
# | Transpose square matrix of size 4096
# | Error at M(0, 18) = 73728
# | FAILED
# `-----------------------------
# error: command failed with exit status: 1
```

The failure must be unrelated to this PR and I'd like @intel/dpcpp-esimd-reviewers to take a look at it.

**Comment by sarnex** [comment]
I have no idea, can you make an internal bug tracker and assign it to our team?

**Pull request merged**